Add long description and test for E0311 (#100747)
rust-lang/rust: compiler/rustc_error_codes/src/error_codes/E0311.md

This error occurs when there is an unsatisfied outlives bound involving an

> elided region on a generic type parameter or associated type.

Reviewer: ```suggestion
elided region and a generic type parameter or associated type.
```

Author: fixed.

This error occurs when there is an unsatisfied outlives bound involving an
elided region on a generic type parameter or associated type.

Erroneous code example:

```compile_fail,E0311
fn no_restriction<T>(x: &()) -> &() {
    with_restriction::<T>(x)
}

fn with_restriction<'a, T: 'a>(x: &'a ()) -> &'a () {
    x
}
```

Why doesn't this code compile? It helps to look at the lifetime bounds that are
automatically added by the compiler. For more details see the documentation for
[lifetime elision]( https://doc.rust-lang.org/reference/lifetime-elision.html).

The compiler elides the lifetime of `x` and the return type to some arbitrary
lifetime `'anon` in `no_restriction()`. The only information available to the
compiler is that `'anon` is valid for the duration of the function. When
calling `with_restriction()`, the compiler requires the completely unrelated

> type parameter `T` to outlive `'anon` because of the `T: 'a bound` in

Reviewer: ```suggestion
type parameter `T` to outlive `'anon` because of the `T: 'a` bound in
```

Reviewer: still relevant

Author: missed this one!  fixed.

`with_restriction()`. This causes an error because `T` is not required to
outlive `'anon` in `no_restriction()`.

If `no_restriction()` were to use `&T` instead of `&()` as an argument, the

> compiler would have added an implied bound [implied

Reviewer 2: ```suggestion
compiler would have added an [implied
```

(Remove the double "implied bound")

Alternatively, remove the link for now, and then send a new PR when the reference PR lands. I don't think it's worth blocking on the reference PR.

Author: I had figured that updating the docs would be quick, but it looks like it's going to be a more involved process.  For now I removed the implied bounds link and will plan to merge this PR. Once https://github.com/rust-lang/reference/pull/1261 has been merged I'll plan to open a smaller PR to add the reference link into the `E0311.md` docs.

> compiler would have added an implied bound [implied
> bound](https://rust-lang.github.io/rfcs/2089-implied-bounds.html), causing this

Reviewer: ```suggestion
compiler would have added an
[implied bound](https://rust-lang.github.io/rfcs/2089-implied-bounds.html), causing this
```
oh wow, we really don't have a user facing doc for implied bounds xx that sucks

the rfc is about implied trait bounds. implied bounds already existed before that and are stable, unlike this rfc. We should really have a page in the reference for that

Reviewer: opened https://github.com/rust-lang/reference/pull/1261 so maybe we should wait for that to land so that you can point there

Author: Thanks for opening the PR to add the reference material.  I added some comments on it and will hold this PR until that one lands.

> E0311 occurs when there is insufficient information for the rust compiler to

Reviewer 3: ```suggestion
This error occurs when there is insufficient information for the rust compiler to
```

> prove that some time has a long enough lifetime.

Reviewer 4: Thank you for the contributions! This explanation could be a bit confusing, I don't know how it should be reworded into, cc @compiler-errors any ideas?

Author: Good point -- it would be great to have a better description here. I'm still pretty new to rust, so this was my initial attempt at reverse engineering why the fix worked.

> to prove that some time has a long enough lifetime.

Reviewer 4: ```suggestion
to prove that a type has a long enough lifetime.
```

Author: fixed.

Erroneous code example:

> ```compile_fail, E0311

Reviewer 3: ``````suggestion
```compile_fail,E0311
``````

use std::borrow::BorrowMut;

trait NestedBorrowMut<U, V> {
    fn nested_borrow_mut(&mut self) -> &mut V;
}

impl<T, U, V> NestedBorrowMut<U, V> for T
where
    T: BorrowMut<U>,

>     U: BorrowMut<V>,  // missing lifetime specifier here --> compile fail

Reviewer 3: ```suggestion
    U: BorrowMut<V>, // error: missing lifetime specifier
```

{
    fn nested_borrow_mut(&mut self) -> &mut V {
        self.borrow_mut().borrow_mut()
    }
}
```

In this example we have a trait that borrows some inner data element of type `V`
from an outer type `T`, through an intermediate type `U`. The compiler is unable

> to prove that the livetime of `U` is long enough to support the reference. To

Reviewer 4: ```suggestion
to prove that the lifetime of `U` is long enough to support the reference. To
```

Author: fixed.

> In this example we have a trait that borrows some inner data element of type V
> from an outer type T, through an intermediate type U. The compiler is unable to
> prove that the livetime of U is long enough to support the reference, so it
> throws E0311. To fix the issue we can explicitly add lifetime specifiers to the
> trait, which link the lifetimes of the various data types and allow the code
> to compile.

Reviewer 3: ```suggestion
In this example we have a trait that borrows some inner data element of type `V`
from an outer type `T`, through an intermediate type `U`. The compiler is unable to
prove that the livetime of `U` is long enough to support the reference. To fix the
issue we can explicitly add lifetime specifiers to the `NestedBorrowMut` trait, which
link the lifetimes of the various data types and allow the code to compile.
```

Working implementation of the `NestedBorrowMut` trait:

```
use std::borrow::BorrowMut;

trait NestedBorrowMut<'a, U, V> {
    fn nested_borrow_mut(& 'a mut self) -> &'a mut V;
}

impl<'a, T, U, V> NestedBorrowMut<'a, U, V> for T
where
    T: BorrowMut<U>,

>     U: BorrowMut<V> + 'a,

Reviewer 3: ```suggestion
    U: BorrowMut<V> + 'a, // Adding lifetime specifier
```

This error occurs when there is an unsatisfied outlives bound involving an
elided region on a generic type parameter or associated type.

Erroneous code example:

```compile_fail,E0311
fn no_restriction<T>(x: &()) -> &() {
    with_restriction::<T>(x)
}

fn with_restriction<'a, T: 'a>(x: &'a ()) -> &'a () {
    x
}
```

Why doesn't this code compile? It helps to look at the lifetime bounds that are
automatically adding by the compiler. For more details see the Rust
Documentation for Lifetime Elision:

> https://doc.rust-lang.org/reference/lifetime-elision.html.

Author: Are there any best practices around how to include links like this in the documentation? For example, would this be better?

```markdown
For more details see the
[Rust Documentation for Lifetime Elision]
(https://doc.rust-lang.org/reference/lifetime-elision.html)
```

There are two lifetimes being passed into the `no_restriction()` function: one
associated with the generic type `T` parameter and the other with the input
argument `x`. The compiler does not know which of these lifetimes can be

> assigned to the output reference, so we get an error.

Author: Let me know if there is a more clear or correct way to explain this concept.

Author: I replaced my description with yours and then applied a few minor changes to smooth out the wording a bit. 

For the reference link, the best I could find is for the RFD on implied bounds. Let me know if you have a better idea.
https://rust-lang.github.io/rfcs/2089-implied-bounds.html

One way to "fix" this code would be to remove the generic type argument `T`.
In this case, the lifetime elision works because there is a single input
lifetime, which is associated with `x`.

```
fn no_restriction(x: &()) -> &() {
    with_restriction(x)
}

fn with_restriction<'a>(x: &'a ()) -> &'a () {
    x
}

> ```

Author: This isn't a true fix, but realizing that this code worked made me realize that the problem had to do with adding the generic parameter in addition to the function argument.

> This error occurs when there is an unsatisfied outlives bound on a generic

Reviewer: ```suggestion
This error occurs when there is an unsatisfied outlives bound involving an elided region on a generic
```
i think that's the difference between E0309 and E0311, maybe we should merge these two error codes 🤷

Author: Thanks for the extra detail here - I updated the text.

        let u_ref = self.borrow_mut();
        let v_ref = u_ref.borrow_mut();
        v_ref
    }
}
```

Why doesn't this code compile? It helps to look at the lifetime bounds that
the compiler is automatically adding ("Lifetime Ellision", Chapter 10.3 in the
Rust book) to the `nested_borrow_mut()` and `borrow_mut()` functions. In both
cases the input is a reference to `self`, so the compiler attempts to assign
the same lifetime to the input and output.

Looking specifically at `nested_borrow_mut()`, we see that there are three
object references to keep track of, along with their associated lifetimes:
- `self` (which is a `&mut T`)
- `u_ref` (which is a `&mut U`)
- `v_ref` (which is a `&mut V`)

The `borrow_mut()` method implicitly requires that that the input and output
have the same lifetime bounds. Thus the lines: `let u_ref = self.borrow_mut();`
and `let v_ref = u_ref.borrow_mut();` in `nested_borrow_mut()` above imply that
`u_ref` and `self` must share a lifetime bound, and also that `v_ref` and
`u_ref` share a lifetime bound. The problem is that the function signature for
`nested_borrow_mut()` only gives the compiler information about the lifetimes

> of `self` and `v_ref` -- nothing about `u_ref`.

Reviewer: i feel like this explanation isn't quite right. `u_ref`, `self` and `v_ref` don't necessarily need the same lifetime, as we may reborrow `self` or `u_ref` for the `borrow_mut` call.

The reason for that error is:
- due to lifetime elision, we have to return a reference with some elided lifetime `'0` from outside of the function.
- because of the signature of `borrow_mut`, `u_ref` has to be valid for at least as long as `v_ref`, meaning that we have some `&'1 V` where `'1: '0` holds.
- this requires `V: '1` which is not implied by the signature (`'1` then ends up being equal to `'0` but that doesn't really matter)

I am not sure whether that's more helpful, but explaining the behavior the following (smaller) example might be easier
```rust
fn no_restriction<T>(x: &()) -> &() {
    with_restriction::<T>(x)
}

fn with_restriction<'a, T: 'a>(x: &'a ()) -> &'a () {
    x
}
```

Author: Thanks for explaining this. I like your minimum working example much better: it makes it more clear exactly what the problem is.  I've replace my original example with your suggestion.

One thing that I missed the first few times that I read this is that the above code works fine if the generic parameter `T` is omitted. I added an example to that effect in the updated documentation because I figure that it might help make it more clear why the lifetime elision fails in this case.

> This error occurs when there is insufficient information for the rust compiler
> to prove that a type has a long enough lifetime.

Reviewer: ```suggestion
This error occurs when there is an unsatisfied outlives bound on a generic type parameter or associated type.
```
something like this maybe?

Author: Yeah - that's much better.

> to prove that a type has a long enough lifetime.

Author: Fixed.

Erroneous code example:

```compile_fail,E0311
use std::borrow::BorrowMut;

trait NestedBorrowMut<U, V> {
    fn nested_borrow_mut(&mut self) -> &mut V;
}

impl<T, U, V> NestedBorrowMut<U, V> for T
where
    T: BorrowMut<U>,
    U: BorrowMut<V>, // error: missing lifetime specifier
{
    fn nested_borrow_mut(&mut self) -> &mut V {
        self.borrow_mut().borrow_mut()
    }
}
```

Why doesn't this code compile? The problem has to do with Rust's rules for
lifetime elision in functions (Chapter 10.3 in the Rust book). One of the
inputs is a reference to `self`, so the compiler attempts to assign the
the same lifetime to the `&mut self` input and `&mut V` output to the
`nested_borrow_mut()` function. The problem is that there is no way for the
compiler to directly figure out how these two lifetimes are related in the
implementation of the function. We're implementing the `NextedBorrowMut`
trait for a type `T`, so the `&mut self` reference has the lifetime of `T`.
We know that `T` implements the `BorrowMut` trait returning a reference to `U`,

> and that `U` implements the `BorrowMut` trait returning a reference to `V`.

Author: I rewrote this to be more clear and correct.

This error occurs when there is an unsatisfied outlives bound involving an
elided region on a generic type parameter or associated type.

Erroneous code example:

```compile_fail,E0311
fn no_restriction<T>(x: &()) -> &() {
    with_restriction::<T>(x)
}

fn with_restriction<'a, T: 'a>(x: &'a ()) -> &'a () {
    x
}
```

Why doesn't this code compile? It helps to look at the lifetime bounds that are
automatically adding by the compiler. For more details see the Rust
Documentation for Lifetime Elision:
https://doc.rust-lang.org/reference/lifetime-elision.html]

There are two lifetimes being passed into the `no_restriction()` function: one
associated with the generic type `T` parameter and the other with the input
argument `x`. The compiler does not know which of these lifetimes can be
assigned to the output reference, so we get an error.

One way to "fix" this code would be to remove the generic type argument `T`.
In this case, the lifetime elision works because there is a single input
lifetime, which is associated with `x`.

```
fn no_restriction(x: &()) -> &() {
    with_restriction(x)
}

fn with_restriction<'a>(x: &'a ()) -> &'a () {
    x
}
```

> The "correct" way to resolve this error is to explicitly tell the compiler
> which input lifetime should be assigned to the output. In this case we give
> both the generic type `T` parameter and the argument `x` the same lifetime
> requirement as the output reference, producing a working version of the code:

Reviewer: ```suggestion
This error can be resolved by eplicitly naming the elided lifetime
and then explicily requiring that the generic parameter outlives that lifetime:
```
